docs/job-configuration-generator.md
---
title: Job Submission Generator
---
<script src="https://code.angularjs.org/1.7.9/angular.js"></script>
<script src="https://code.angularjs.org/1.7.9/angular-messages.js"></script>

!!!info "Automatic parameter selection"
    - You can manually specify parameters at job submission using the command below. If needed, all parameters [can also be automatically configured at queue level](../tutorials/integration-ec2-job-parameters/#how-to-use-custom-parameters). 
    - Job will use the default parameters configured for its queue unless the parameters are explicitly specified during submission (**job parameters override queue parameters**).
    - [Refer to this page](../tutorials/launch-your-first-job/#examples) for additional examples.
    

<body ng-app="myApp">
<div ng-controller="myCtrl">

<style>
* {
  box-sizing: border-box;
}

.input2 {
  padding: 12px;
  width: 85%;
  border: 1px solid #ccc;
  border-radius: 4px;
  resize: vertical;
  font-size: 15px;
  margin-top: 6px;
}

.container {
  border-radius: 5px;
  background-color: #f2f2f2;
  padding: 5px;
}

.col-25 {
  float: left;
  width: 25%;
  margin-top: 6px;
}

.col-75 {
  float: left;
  width: 75%;
  margin-top: 6px;
}

/* Responsive layout - when the screen is less than 600px wide, make the two columns stack on top of each other instead of next to each other */
@media screen and (max-width: 600px) {
  .col-25, .col-75, input[type=submit] {
    width: 100%;
    margin-top: 0;
  }
}


.md-content {
margin-right: 0;
}

</style>






<div class="container">



<h2> Your Command </h2>
<strong>user@host:</strong> qsub {{qsub_instance_ami}} 
{{qsub_instance_type}} 
{{qsub_subnet_id}} 
{{qsub_spot_price}}
{{qsub_efa_support}}
{{qsub_placement_group}}
{{qsub_root_size}}
{{qsub_scratch_size}}
{{qsub_scratch_iops}}
{{qsub_fsx_lustre}}
{{qsub_fsx_lustre_size}}
{{qsub_ht_support}}
{{qsub_spot_allocation_count}}
{{qsub_spot_allocation_strategy}}
{{qsub_nodes}}
{{qsub_base_os}}
{{qsub_keep_ebs}}
{{qsub_force_ri}}
myscript.sh

<h2> Job Parameters</h2>

<form name="QsubForm">

<h3> Compute parameters: </h3>
    
   <input class="input2" min="1" name="nodes" size="35" type="text" ng-change="myFunc()" ng-model="nodes" placeholder="How many EC2 instance(s) do you want to provision for your job?" pattern="^[1-9][0-9]*$"/>
          <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#nodes">Documentation</a></i>
        
   <div ng-messages="QsubForm.nodes.$error">
            <div style="color: red; font-size: medium" ng-message="pattern">Must be a number greater than 0</div>
   </div>
   
   <input class="input2"  size="35" style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="instance_type" placeholder="What type of instance do you need? (use + to specify more than one eg. c5.large+m5.large)" />
   <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#instance_type">Documentation</a></i>
   

   <input class="input2" required name="instance_ami" size="35" ng-minlength="3" style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="instance_ami" placeholder="Enter the AMI ID if you are planning to use a custom AMI" pattern="^ami-[a-zA-Z0-9]*" />
   <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#instance_ami">Documentation</a></i>
   <div ng-messages="QsubForm.instance_ami.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Image name must start with "ami-"</div>
   </div>
   
   <input class="input2" required name="base_os" size="35" ng-minlength="3" style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="base_os" placeholder="What's the operating system of your AMI (if different that the base os specified during install)" pattern="centos7|rhel7|amazonlinux2" />
   <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#base_os">Documentation</a></i>

   <div ng-messages="QsubForm.base_os.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Must be centos7, rhel7 or amazonlinux2</div>
   </div>
      <div style="color: red; font-size: medium">{{base_os_error}}</div>

   
   
   <input class="input2"  name="subnet_id" size="35" style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="subnet_id" placeholder="Enter the subnet ID if you are planning to use a specific subnet" pattern="^sub-[a-zA-Z0-9]*" />
      <i><a style="padding: 5px"target="_blank"  href="../tutorials/integration-ec2-job-parameters/#subnet_id">Documentation</a></i>
    <div ng-messages="QsubForm.subnet_id.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Subnet name must start with "sub-"</div>
   </div>
   
   <input class="input2"  name="spot_price" size="35" style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="spot_price" placeholder="Need SPOT instances? Enter the maximum price you are willing to pay (or auto to match OD price)" pattern="[+-]?([0-9]*[.])?[0-9]+|auto"/>
   <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#spot_price">Documentation</a></i>

   <div ng-messages="QsubForm.spot_price.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Spot Price must be a float (eg 1.2) or auto (match OD price)</div>
   </div>
   
   <input class="input2"  name="spot_allocation_count" size="35" style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="spot_allocation_count" placeholder="Want to launch both OD and SPOT? Specify the number of SPOT instances to provision for your job"  pattern="^[1-9][0-9]*$" />
   <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#spot_allocation_count">Documentation</a></i>

   <div ng-messages="QsubForm.spot_allocation_count.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Must be a number</div>
   </div>
   <div style="color: red; font-size: medium">{{spot_allocation_error}}</div>
   <div style="color: red; font-size: medium">{{spot_allocation_error_price}}</div>

   <input class="input2"  name="spot_allocation_strategy" size="35" style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="spot_allocation_strategy" placeholder="What's your SPOT allocation strategy when using more than 1 SPOT instance type" pattern="lowest-cost|capacity-optimized"/>
      <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#spot_allocation_strategy">Documentation</a></i>
   <div ng-messages="QsubForm.spot_allocation_strategy.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Must be either lowest-cost (default) or capacity-optimized</div>
   </div>
    
   <div style="color: red; font-size: medium">{{spot_allocation_strategy_price}}</div>



<h3> Storage parameters: </h3>   
   
   
   <input class="input2"  name="root_size" size="35"  style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="root_size" placeholder="How much storage to allocate for the root partition (in GB)"  pattern="\d+" />
      <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#root_size">Documentation</a></i>

   <div ng-messages="QsubForm.root_size.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Root Size must be a number</div>
   </div>
   
   <input class="input2"  name="scratch_size" size="35"  style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="scratch_size" placeholder="How much storage to allocate for the scratch partition (in GB)" pattern="\d+"/>
      <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#scratch_size">Documentation</a></i>

   <div ng-messages="QsubForm.scratch_size.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Scratch Size must be a number</div>
   </div>
   
  
   <input class="input2"  size="50" name = "scratch_iops" style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="scratch_iops" placeholder="Do you need IOPS optimized EBS for /scratch? If yes enter the number of provisioned IO to provision" pattern="\d+" />
       <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#scratch_iops">Documentation</a></i>

   <div ng-messages="QsubForm.scratch_iops.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Provisioned IO/s must be a number</div>
   </div>
   
   
   <input class="input2"  size="35"  style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="fsx_lustre" placeholder="Need FSx for Lustre? User either True, a S3 bucket name or an existing Filesystem ID" />
      <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#fsx_lustre">Documentation</a></i>

   <input class="input2"  name="fsx_lustre_size" size="35"  style="font-size: 15px;  margin-top: 6px;" type="text" ng-change="myFunc()" ng-model="fsx_lustre_size" placeholder="How much storage do you want to provision for your FSx partition (in GB)" pattern="\d+" />
      <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#fsx_lustre_size">Documentation</a></i>
      <div ng-messages="QsubForm.fsx_lustre_size.$error">
        <div style="color: red; font-size: medium" ng-message="pattern">Size must be a number</div>
   </div>
   
<h3>Flags: </h3>
   <input type="checkbox" ng-change="myFunc()" ng-model="efa_support"> I want to use EFA <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#efa_support">Documentation</a></i> <br>
   <input type="checkbox" ng-change="myFunc()" ng-model="placement_group"> I do not want to use Placement Group (enabled by default)    <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#placement_group">Documentation</a></i> <br>
   <input type="checkbox" ng-change="myFunc()" ng-model="ht_support"> I want to enable HyperThreading (disabled by default)   <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#ht_support">Documentation</a></i><br>
   <input type="checkbox" ng-change="myFunc()" ng-model="keep_ebs"> I want to retain my EBS disks (disabled by default)   <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#keep_ebs">Documentation</a></i><br>
   <input type="checkbox" ng-change="myFunc()" ng-model="force_ri"> I want my job to only run on Reserved instances  <i><a style="padding: 5px" target="_blank" href="../tutorials/integration-ec2-job-parameters/#force_ri">Documentation</a></i><br>

</form> 
  

</div>

<script>
  angular.module('myApp', ['ngMessages'])
    .controller('myCtrl', ['$scope', function($scope) {
      $scope.count = 0;
      $scope.myFunc = function() {
        if($scope.nodes){$scope.qsub_nodes = "-l nodes=" + $scope.nodes;}else{$scope.qsub_nodes = "";}
        if($scope.instance_ami){$scope.qsub_instance_ami = "-l instance_ami=" + $scope.instance_ami;}else{$scope.qsub_instance_ami = "";}
        if($scope.base_os){$scope.qsub_base_os = "-l base_os=" + $scope.base_os;}else{$scope.qsub_base_os = "";}
        if($scope.instance_type){$scope.qsub_instance_type = "-l instance_type=" + $scope.instance_type;}else{$scope.qsub_instance_type = "";}
        if($scope.subnet_id){$scope.qsub_subnet_id = "-l subnet_id=" + $scope.subnet_id;}else{$scope.qsub_subnet_id = "";}
        if($scope.spot_price){$scope.qsub_spot_price = "-l spot_price=" + $scope.spot_price;}else{$scope.qsub_spot_price= "";}
        if($scope.root_size){$scope.qsub_root_size= "-l root_size=" + $scope.root_size;}else{$scope.qsub_root_size= "";}
        if($scope.scratch_size){$scope.qsub_scratch_size = "-l scratch_size=" + $scope.scratch_size;}else{$scope.qsub_scratch_size= "";}
        if($scope.scratch_iops){$scope.qsub_scratch_iops= "-l scratch_iops=" + $scope.scratch_iops;}else{$scope.qsub_scratch_iops= "";}     
        if($scope.spot_allocation_count){$scope.qsub_spot_allocation_count= "-l spot_allocation_count=" + $scope.spot_allocation_count;}else{$scope.qsub_spot_allocation_count= "";}     
        if($scope.spot_allocation_strategy){$scope.qsub_spot_allocation_strategy= "-l spot_allocation_strategy=" + $scope.spot_allocation_strategy;}else{$scope.qsub_spot_allocation_strategy= "";}      
        if($scope.keep_ebs){$scope.qsub_keep_ebs = "-l keep_ebs=True";}else{$scope.qsub_keep_ebs= "";}
        if($scope.efa_support){$scope.qsub_efa_support = "-l efa_support=True";}else{$scope.qsub_efa_support= "";}
        if($scope.placement_group){$scope.qsub_placement_group = "-l placement_group=False";}else{$scope.qsub_placement_group= "";}
        if($scope.ht_support){$scope.qsub_ht_support = "-l ht_support=True";}else{$scope.qsub_ht_support= "";}
        if($scope.fsx_lustre){$scope.qsub_fsx_lustre = "-l fsx_lustre=" + $scope.fsx_lustre;}else{$scope.qsub_fsx_lustre = "";}
        if($scope.fsx_lustre_size){$scope.qsub_fsx_lustre_size = "-l fsx_lustre_size=" + $scope.fsx_lustre_size;}else{$scope.qsub_fsx_lustre_size = "";}
         if($scope.force_ri){$scope.qsub_force_ri = "-l force_ri=True";}else{$scope.qsub_fsx_lustre = "";}

        if (!$scope.nodes){$scope.nodes_count=0;}else{$scope.nodes_count=$scope.nodes}
            
        
        if (+$scope.spot_allocation_count >= +$scope.nodes_count)
        {
            $scope.spot_allocation_error = "Allocated spots must be lower than the number of nodes provisioned for your job";
            $scope.qsub_spot_allocation_count= "";
        }
        else
        {
         $scope.spot_allocation_error = "";
        }
        
        
        if ($scope.spot_allocation_count && !$scope.spot_price)
        {
            $scope.spot_allocation_error_price = "Spot Price must be specified";
            $scope.qsub_spot_allocation_count= "";
        }
        else
        {
         $scope.spot_allocation_error_price = "";
        }
        
         if ($scope.spot_allocation_strategy && !$scope.spot_price)
        {
            $scope.spot_allocation_strategy_price = "Spot Price must be specified";
            $scope.qsub_spot_allocation_strategy= "";
        }
        else
        {
         $scope.spot_allocation_strategy_price = "";
        }
        
        
        if ($scope.base_os && !$scope.instance_ami)
        {
            $scope.base_os_error = "No need to specify base_os if you don't use a custom AMI";
            $scope.qsub_base_os= "";
        }
        else
        {
         $scope.base_os_error = "";
        }
         
 
      
      };
    }]);
</script>




</body>



docs/storage/backend-storage-options.md
---
title: Understand backend storage
---

!!! info "Scale-Out Computing on AWS gives you the flexibility to customize your storage backend based on your requirements"
    - You can customize the root partition size
    - You can provision a local scratch partition
    - You can deploy standard SSD (gp2) or IO Optimized SSD (io1) volumes
    - Scale-Out Computing on AWS automatically leverages instance store disk(s) as scratch partition when applicable
    - In term of performance: Instance Store > EBS SSD IO > EBS SSD Standard > EFS
    
    [Refer to this link](https://aws.amazon.com/ebs/features/) to learn more about EBS volumes.


## EFS (shared) partitions

### /data partition

/data is an [Elastic File System](https://aws.amazon.com/efs/) partition mounted on all hosts. This contains the home directory of your LDAP users ($HOME = `/data/home/<USERNAME>`). **This partition is persistent**. Avoid using this partition if your simulation is disk I/O intensive (use /scratch instead)

### /apps partition

/apps is an [Elastic File System](https://aws.amazon.com/efs/) partition mounted on all hosts. This partition is designed to host all your CFD/FEA/EDA/Mathematical applications. **This partition is persistent**. Avoid using this partition if your simulation is disk I/O intensive (use /scratch instead)

## FSx

Scale-Out Computing on AWS supports FSx natively. [Click here to learn how to use FSx as backend storage for your jobs](../../storage/launch-job-with-fsx/).

## Instance (local) partitions

Below are the storage options you can configure at an instance level for your jobs. If needed, add/remove/modify the storage logic by editing `ComputeNode.sh` script to match your requirements.

### Root partition

By default Scale-Out Computing on AWS provision a 10GB EBS disk for the root partition. This may be an issue if you are using a custom AMI configured with a bigger root disk size or if you simply want to allocate additional storage for the root partition. 
To expand the size of the volume, submit a simulation using `-l root_size=<SIZE_IN_GB>` parameter.

~~~bash
user@host:qsub -l root_size=25 -- /bin/sleep 600
~~~

Result: Root partition is now 25GB

~~~bash hl_lines="3 9"
user@host: lsblk
NAME          MAJ:MIN RM SIZE RO TYPE MOUNTPOINT
nvme0n1       259:0    0  25G  0 disk
├─nvme0n1p1   259:1    0  25G  0 part /
└─nvme0n1p128 259:2    0   1M  0 part

user@host: df -h /
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme0n1p1   25G  2.2G   23G   9% /
~~~

### Scratch Partition

!!! info
    - It's recommended to provision `/scratch` directory whenever your simulation is I/O intensive.
    - `/scratch` is a local partition and will be deleted when you job complete. Make sure to copy the job output back to your $HOME
    - `/scratch` is automatically created when Instance supports local ephemeral storage



#### Request a /scratch partition with SSD disk

During job submission, specify `-l scratch_size=<SIZE_IN_GB>` to provision a new EBS disk (`/dev/sdj`) mounted as `/scratch`

~~~bash
user@host: qsub -l scratch_size=150 -- /bin/sleep 600
~~~

Result: a 150 GB /scratch partition is available on all nodes

~~~bash hl_lines="3 10"
user@host: lsblk
NAME          MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
nvme1n1       259:0    0  150G  0 disk /scratch
nvme0n1       259:1    0   10G  0 disk
├─nvme0n1p1   259:2    0   10G  0 part /
└─nvme0n1p128 259:3    0    1M  0 part

user@host: df -h /scratch
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme1n1    148G   61M  140G   1% /scratch
~~~

To verify the type of your EBS disk, simply go to your AWS bash > EC2 > Volumes and verify your EBS type is "gp2" (SSD). 
[Refer to this link](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html) for more information about the various EBS types available.

![](../imgs/storage-1.png)

#### Request a /scratch partition with IO optimized disk

To request an optimized SSD disk, use `-l scratch_iops=<IOPS>` along with `-l scratch_size=<SIZE_IN_GB>`. [Refer to this link](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html) to get more details about burstable/IO EBS disks.
~~~bash
user@host: qsub -l scratch_iops=6000 -l scratch_size=200 -- /bin/sleep 600
~~~

Looking at the EBS bash, the disk type is now "io1" and the number of IOPS match the value specified at job submission.

![](../imgs/storage-2.png)

### Instance store partition

!!! warning "Free storage is always good"
    You are not charged for instance storage (included in the price of the instance)


[Some instances come with default instance storage](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/InstanceStorage.html).
An instance store provides temporary block-level storage for your instance. This storage is located on disks that are physically attached to the host computer and is removed as soon as the node is deleted.

Scale-Out Computing on AWS automatically detects instance store disk and will use them as /scratch **unless you specify `-l scratch_size` parameter for your job**. In this case, Scale-Out Computing on AWS honors the user request and ignore the instance store volume(s).

#### When node has 1 instance store volume

For this example, I will use a "c5d.9xlarge" instance which is coming with a 900GB instance store disk.

~~~bash
user@host: qsub -l instance_type=c5d.9xlarge -- /bin/sleep 600
~~~

Result: Default /scratch partition has been provisioned automatically using local instance storage

~~~bash hl_lines="3 10"
user@host: lsblk
NAME          MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
nvme1n1       259:0    0 838.2G  0 disk /scratch
nvme0n1       259:1    0    10G  0 disk
├─nvme0n1p1   259:2    0    10G  0 part /
└─nvme0n1p128 259:3    0     1M  0 part

user@host: df -h /scratch
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme1n1    825G   77M  783G   1% /scratch
~~~


#### When node has more than 1 instance store volumes

In this special case, `ComputeNode.sh` script will create a raid0 partition using all instance store volumes available.
 
For this example, I will use a "m5dn.12xlarge" instance which is shipped with a 2 * 900GB instance store disks (total 1.8Tb).

~~~bash
user@host: qsub -l instance_type=m5dn.12xlarge -- /bin/sleep 600
~~~

Result: /scratch is a 1.7TB raid0 partition (using 2 instance store volumes)

~~~bash hl_lines="3 5 13"
user@host: lsblk
NAME          MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
nvme1n1       259:0    0 838.2G  0 disk
└─md127         9:127  0   1.7T  0 raid0 /scratch
nvme2n1       259:1    0 838.2G  0 disk
└─md127         9:127  0   1.7T  0 raid0 /scratch
nvme0n1       259:2    0    10G  0 disk
├─nvme0n1p1   259:3    0    10G  0 part  /
└─nvme0n1p128 259:4    0     1M  0 part

user@host: df -h /scratch
Filesystem      Size  Used Avail Use% Mounted on
/dev/md127      1.7T   77M  1.6T   1% /scratch
~~~
### Combine custom scratch and root size

You can combine parameters as needed. For example, `qsub -l root_size=150 -l scratch_size=200 -l nodes=2` will provision 2 nodes with 150GB / and 200GB SSD /scratch

### Change the storage parameters at queue level

Edit `/apps/soca/<CLUSTER_ID>/cluster_manager/settings/queue_mapping.yml` to configure default storage settings at a queue level:

~~~yaml hl_lines="7 8 15 16 23"
queue_type:
  compute:
    # /root will be 30 GB and /scratch will be a standard 100GB SSD
    queues: ["queue1", "queue2", "queue3"]
    instance_ami: "ami-082b5a644766e0e6f"
    instance_type: "c5.large"
    scratch_size: "100"
    root_size: "30"
    # .. Refer to the doc for more supported parameters
  memory:
    # /scratch will be a SSD with provisioned IO
    queues: ["queue4"]
    instance_ami: "ami-082b5a644766e0e6f"
    instance_type: "r5.large"
    scratch_size: "300"
    scratch_iops: "5000"

  instancestore:
    # /scratch will use the default instance store
    queues: ["queue5"]
    instance_ami: "ami-082b5a644766e0e6f"
    instance_type: "m5dn.12large"
    root_size: "300"
~~~


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.5.0] - 2020-07-17
### Added
- Support for Elastic MetricBeat
- Added HTTP REST API to interact with SOCA
- Users can now decide to restrict a job to Reserved Instances
- Revamped Web Interface
  - Added filesystem explorer
  - Users can upload files/folders via drag & drop interface
  - Users can edit files directly on SOCA using a cloud text editor
  - Users can now manage membership of their own LDAP group via web
  - Users can now understand why they job is not started  (eg: instance issue, misconfiguration, AWS limit, license limit) directly on the UI
  - Users can kill their job via the web
  - Admins can manage SOCA LDAP via web (create group, user, manage ownership and permissions)
  - Admins can creates application profiles and let user submit job via web interface
  - Ability to trigger Linux commands via HTML form
- Admins can now limit the number of running jobs per queue
- Admins can now limit the number of running instances per queue
- Admins can now specify the idle timeout value for any DCV sessions. Inactive DCV sessions will be automatically terminated after this period
- Job selection can now configured at queue level (FIFO or fair share)
- Dry run now supports vCpus limit
- Support for custom shells

### Changed
- Updated Troposphere to 2.6.1
- Updated EFA to 1.9.3
- Updated Nice DCV to 2020.0-8428
- Updated ElasticSearch to 7.4
- You can specify a name for your DCV sessions 
- You can now specify custom AMI, base OS or storage options for your DCV sessions
- Project assigned to DCV jobs has been renamed to "remotedesktop" (previously "gui")
- Dispatcher script is now running every minute
- SOCA now deploys 2 instances for ElasticSearch for high availability
- Users can now specify DEPLOYMENT_TYPE for their FSX for Lustre filesystems
- Users can specify PerUnitThroughput when FSx for Lustre deployment type is set to PERSISTENT
- DCV now supports G4 instance type (#24)
- X11 is now configured correctly for ALI 3D DCV session (#23)


## [2.0.1] - 2020-04-20
### Added
- Support for SpotFleet

### Changed
- NVIDIA drivers are now automatically installed when a GPU instance is provisioned
- Deployed MATE Desktop for DCV for Amazon Linux 2

## [2.0.0] - 2020-03-18
### Added

- Support for MixedInstancePolicy and InstanceDistribution
- Support for non-EBS optimized instances such as t2
- Integration of AWS Session Manager
- Integration of AWS Backup
- Integration of AWS Cognito
- Integration of Troposphere
- Admins can now manage ACL (individual/LDAP groups) at queue level
- Admins can now restrict specific type/family of instance at queue level
- Admins can now prevent users to change specific EC2 parameters
- Users can now install SOCA using existing resources such as VPC, Security Groups ...
- Users now have the ability to retain EBS disks associated to a simulation for debugging purposes
- SOCA now prevent jobs to be submitted if .yaml configuration files are malformed
- Scheduler Root EBS is now tagged with cluster ID 
- Scheduler Network Interface is now tagged with cluster ID 
- Scheduler and Compute hosts are now sync with Chrony (Amazon Time Sync)
- Support for FSx for Lustre new Scratch2/Scratch1 and Persistent mode
- Added Compute nodes logs on EFS (/apps/soca/<cluster_id>/cluster_node_bootstrap/logs/<job_id>/<host>/*.log) for easy debugging

### Changed

- Ignore installation if PBSPro is already configured on the AMI
- Fixed bug when stack name only use uppercase
- ComputeNode bootstrap scripts are now loaded from EFS
- Users can now open a SSH session using SSM Session Manager
- Processes are now automatically launched upon scheduler reboot 
- Spot price now default to the OD price
- Default admin password now supports special characters
- Ulimit is now disabled by default on all compute nodes
- Dispatcher automatically append "s3://" if not present when using FSx For Lustre
- Updated default ElasticSeach instance to m5.large to support encryption at rest
- SOCA libraries are now installed under /apps/soca/<CLUSTER_ID> location to support multi SOCA environments 
- Web UI now display the reason when a DCV job can't be submitted
- Customers can now provision large number of EC2 hosts accross multiple subnets using a single API call 
- Smart detection of Placement Group requirement when using more than 1 subnet
- Added retry mechanism for some AWS API calls which throttled when provisioning > 1000 nodes in a single API call
- ALB Target Groups are now correctly deleted once the DCV sessions is terminated
- SOCA version is now displayed on the web interface
- Updated EFA version to 1.8.3

## [1.0.0] - 2019-11-20
- Release Candidate



source/soca/cluster_web_ui/config.py
import os
from datetime import timedelta
import secrets
from cryptography.fernet import Fernet

basedir = os.path.abspath(os.path.dirname(__file__))


class Config(object):
    # APP
    DEBUG = False
    TESTING = False
    USE_PERMANENT_SESSION = True
    PERMANENT_SESSION_LIFETIME = timedelta(days=1)
    SESSION_TYPE = "sqlalchemy"
    SQLALCHEMY_TRACK_MODIFICATIONS = False
    SESSION_SQLALCHEMY_TABLE = "flask_sessions"
    SQLALCHEMY_DATABASE_URI = "sqlite:///" + os.path.join(basedir, "db.sqlite")
    SECRET_KEY = os.environ["SOCA_FLASK_SECRET_KEY"]
    API_ROOT_KEY = os.environ["SOCA_FLASK_API_ROOT_KEY"]
    SOCA_DATA_SHARING_SYMMETRIC_KEY = os.environ["SOCA_FLASK_FERNET_KEY"]

    # WEB
    APPS_LOCATION = "/apps/"
    USER_HOME = "/data/home"
    CHROOT_USER = False  # if True, user can only access their $HOME directory (aka: USER_HOME/<user>)
    PATH_TO_RESTRICT = ['/bin', '/boot', '/dev', '/etc', '/home', '/lib', '/lib64', '/local',
                        '/media', '/opt', '/proc', '/root', '/run', '/sbin', '/srv', '/sys', '/tmp', '/usr',
                        '/var']  # List of folders not accessible via the web ui
    DEFAULT_CACHE_TIME = 120  # 2 minutes. Change this value to optimize performance in case you have a large number of concurrent user
    MAX_UPLOAD_FILE = 5120  # 5 GB
    MAX_UPLOAD_TIMEOUT = 1800000  # 30 minutes
    MAX_SIZE_ONLINE_PREVIEW = 150000000  # in bytes (150mb by default), maximum size of file that can be visualized via the web editor
    MAX_ARCHIVE_SIZE = 150000000  # in bytes (150mb by default), maximum size of archive generated when downloading multiple files at once
    DAILY_BACKUP_COUNT = 15  # Keep 15 latest daily backups
    KIBANA_JOB_INDEX = "job*"  # Default index to look for /my_activity. Change it something more specific if using more than 1 index with name ~ "job*"

    # UWSGI SETTINGS
    FLASK_HOST = "127.0.0.1"
    FLASK_PROTOCOL = "https://"
    FLASK_PORT = "8443"
    FLASK_ENDPOINT = FLASK_PROTOCOL + FLASK_HOST + ":" + FLASK_PORT

    # COGNITO
    ENABLE_SSO = False
    COGNITO_OAUTH_AUTHORIZE_ENDPOINT = "https://<YOUR_COGNITO_POOL>.auth.<YOUR_REGION>.amazoncognito.com/oauth2/authorize"
    COGNITO_OAUTH_TOKEN_ENDPOINT = "https://<YOUR_COGNITO_POOL>.auth.<YOUR_REGION>.amazoncognito.com/oauth2/token"
    COGNITO_JWS_KEYS_ENDPOINT = "https://cognito-idp.<YOUR_REGION>.amazonaws.com/<YOUR_REGION>_<YOUR_ID>/.well-known/jwks.json"
    COGNITO_APP_SECRET = "<YOUR_APP_SECRET>"
    COGNITO_APP_ID = "<YOUR_APP_ID>"
    COGNITO_ROOT_URL = "<YOUR_WEB_URL>"
    COGNITO_CALLBACK_URL= "<YOUR_CALLBACK_URL>"

    # DCV
    DCV_AUTH_DIR = "/var/run/dcvsimpleextauth"
    DCV_SIMPLE_AUTH = "/usr/libexec/dcvsimpleextauth.py"
    DCV_SESSION_LOCATION = "tmp/dcv_sessions"
    DCV_MAX_SESSION_COUNT = 4
    DCV_TERMINATE_IDLE_SESSION = 0  # In hours. DCV session will be terminated if there is no active connection within the time specified. 0 to disable

    # LDAP
    LDAP_HOST = "127.0.0.1"
    LDAP_BASE_DN = "dc=soca,dc=local"
    LDAP_ADMIN_PASSWORD_FILE = "/root/OpenLdapAdminPassword.txt"
    LDAP_ADMIN_USERNAME_FILE = "/root/OpenLdapAdminUsername.txt"
    ROOT_DN = 'CN=' + open(LDAP_ADMIN_USERNAME_FILE, 'r').read().rstrip().lstrip() + ',' + LDAP_BASE_DN
    ROOT_PW = open(LDAP_ADMIN_PASSWORD_FILE, 'r').read().rstrip().lstrip()

    # PBS
    PBS_QSTAT = "/opt/pbs/bin/qstat"
    PBS_QDEL = "/opt/pbs/bin/qdel"
    PBS_QSUB = "/opt/pbs/bin/qsub"
    PBS_QMGR = "/opt/pbs/bin/qmgr"

    # SSH
    SSH_PRIVATE_KEY_LOCATION = "tmp/ssh"


app_config = Config()